### PR TITLE
PIM-2455: Remove common attributes used in variant groups

### PR DESCRIPTION
**Pull request by sumbobyboys**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | Y 
| BC breaks?           | N
| CI currently passes? | Y
| Tests pass?          | Y
| Scenarios pass?      | ?
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | N
| Fixed tickets        | PIM-2455
| DB schema updated?   | N
| Migration script?    | N
| Doc PR               |

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php [hunk 1]
```diff
@@ -260,22 +283,109 @@ public function initialize()
...
+        }
+
+        return $variantAttributes;
+    }
+
+    /**
+     * Get warning messages to display during the mass edit action
+     * @return array
+     */
+    protected function generateWarningMessages()
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php]: really sad about this method but as long as we have this shitty `initialize` method, we cannot generate the warning messages "on the fly" :/

File: src/Pim/Bundle/EnrichBundle/Resources/views/MassEditAction/configure/edit-common-attributes.html.twig [hunk 2]
```diff
@@ -38,19 +49,7 @@
         </div>
     {% endfor %}
 
-    {% set attributeCount = form.operation.vars.data.commonAttributes|length %}
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/views/MassEditAction/configure/edit-common-attributes.html.twig]: We should really avoid to do that in the past

File: src/Pim/Bundle/EnrichBundle/Resources/views/MassEditAction/configure/edit-common-attributes.html.twig [hunk 3]
```diff
@@ -38,19 +49,7 @@
...
-    {% if attributeCount == 0 %}
-        {% set alertClass = 'info' %}
-        {% set alertMessage = 'pim_enrich.mass_edit_action.edit-common-attributes.message.warning'|trans %}
-    {% endif %}
 
-    {% if alertClass is defined and alertMessage is defined %}
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/views/MassEditAction/configure/edit-common-attributes.html.twig]: This kind of test is kind of crazy for example

File: src/Pim/Bundle/EnrichBundle/Resources/views/MassEditAction/configure/edit-common-attributes.html.twig [hunk 4]
```diff
@@ -38,19 +49,7 @@
         </div>
     {% endfor %}
 
-    {% set attributeCount = form.operation.vars.data.commonAttributes|length %}
-    {% if attributeCount == 0 %}
-        {% set alertClass = 'info' %}
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/views/MassEditAction/configure/edit-common-attributes.html.twig]: not extensibility...

File: src/Pim/Bundle/CatalogBundle/Entity/ProductTemplate.php [hunk 5]
```diff
@@ -77,4 +77,14 @@ public function hasValueForAttribute(AttributeInterface $attribute)
     {
         return isset($this->valuesData[$attribute->getCode()]);
     }
+
+    /**
+     * Get attributes of the product template
+     * TODO: this method could be optimized by storing the attributes appart in db
```
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Entity/ProductTemplate.php]: yes could be interesting, depends on use case, we'll see

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php [hunk 6]
```diff
@@ -244,7 +257,17 @@ public function initialize()
...
         }
-        $this->initializeCommonAttributes($productIds);
+
+        // Set attribute options locale
+        $currentLocaleCode = $this->getLocale()->getCode();
+        $this->catalogContext->setLocaleCode($currentLocaleCode);
+
+        $attributes = $this->generateCommonAttributes($productIds, $currentLocaleCode);
+        $attributes = $this->filterLocaleSpecificAttributes($attributes, $currentLocaleCode);
+        $attributes = $this->filterAttributesCommingFromVariant($attributes, $this->objects);
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php]: filterAttributesCommingFromVariant -> filterAttributesComingFromVariant

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php [hunk 7]
```diff
@@ -260,22 +283,109 @@ public function initialize()
...
+    }
+
+    /**
+     * Filter attribute by removing attributes coming from variants
+     * @param AttributeInterface[] $attributes
+     * @param array $products
+     *
+     * @return AttributeInterface[]
+     */
+    protected function filterAttributesCommingFromVariant(array $attributes, array $products)
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php]: Comming -> Coming, not sure about this naming, perhaps use another word than coming, belonging or something like that \cc @filipsalpe your opinion ?

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php [hunk 8]
```diff
@@ -260,22 +283,109 @@ public function initialize()
...
+
+    /**
+     * Get warning messages to display during the mass edit action
+     * @return array
+     */
+    protected function generateWarningMessages()
+    {
+        $messages = [];
+
+        $variantAttributes = $this->getCommonVariantAttributes();
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php]: generateCommonAttributes or getCommonAttributes ?

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php [hunk 9]
```diff
@@ -260,22 +283,109 @@ public function initialize()
      *   - is not common to every products
      *
      * @param array $productIds
+     *
+     * @return array
      */
-    protected function initializeCommonAttributes(array $productIds)
+    protected function generateCommonAttributes(array $productIds, $currentLocaleCode)
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php]: please re-order public and protected methods

**Comment by nidup** [comment]
I would extract the whole computation of common attributes in another service, for instance a CommonAttributesProvider/Repository/Resolver and inject this service to the EditCommonAttributes, then in EditCommonAttributes just do a $myOtherService->getCommonAttributes(productIds);

File: src/Pim/Bundle/CatalogBundle/Manager/ProductMassActionManager.php [hunk 10]
```diff
@@ -47,19 +47,107 @@ public function __construct(
      *
      * @param array $productIds
```
Comment (solivier) [inline src/Pim/Bundle/CatalogBundle/Manager/ProductMassActionManager.php]: update php doc

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php [hunk 11]
```diff
@@ -344,59 +357,28 @@ protected function setProductValues(ProductInterface $product)
      *
      * @param AttributeInterface $attribute
```
Comment (solivier) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php]: update php doc

File: features/Context/DataGridContext.php [hunk 12]
```diff
@@ -662,6 +662,7 @@ public function iPutTheColumnBeforeTheOne($source, $target)
      */
     public function iMassEditEntities($entities)
     {
+        $this->wait();
```
Comment (filipsalpe) [inline features/Context/DataGridContext.php]: Is this required?

Comment (sumbobyboys) [inline features/Context/DataGridContext.php]: nope, sorry about that

File: src/Pim/Bundle/EnrichBundle/Resources/translations/messages.en.yml [hunk 13]
```diff
@@ -723,8 +723,8 @@ pim_enrich.mass_edit_action:
         description:   The selected product's attributes will be edited with following data for the chosen locale.
         empty:         Please select the attribute(s) you want to edit
         success_flash: Product(s) attribute(s) have been updated
-        message:
-            warning: There is no attribute in common for the selected products. Please change your selection.
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/Resources/translations/messages.en.yml]: please never delete any translation key

File: spec/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributesSpec.php [hunk 14]
```diff
@@ -24,13 +26,11 @@
 class EditCommonAttributesSpec extends ObjectBehavior
 {
     function let(
-        ProductManager $productManager,
+        ProductBuilder $productBuilder,
```
Comment (nidup) [inline spec/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributesSpec.php]: means a missing BC Break in Changelog, we already have a line for this class imho but we should complete it

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php [hunk 15]
```diff
@@ -80,27 +73,23 @@ class EditCommonAttributes extends ProductMassEditOperation
      * @param array                    $classes
      */
     public function __construct(
-        ProductManager $productManager,
+        ProductBuilder $productBuilder,
```
Comment (nidup) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/EditCommonAttributes.php]: BC break is missing in changelog

**Pull request merged**